rrs/views.py

```diff
@@ -0,0 +1,1005 @@
+                        WHERE status = 'N'
+                        AND history_id = %s
+                    """, [date_id])
+        return [i[0] for i in cur.fetchall()]
+
+    @staticmethod
+    def get_reupg_by_date(layerbranch_id, date):
+        """ Get info for Recipes for the milestone """
+        cur = connection.cursor()
+        cur.execute("""SELECT rs.id, rs.pn, rs.summary, te.version, rownum FROM (
+                            SELECT recipesymbol_id, version, commit_date, upgrade_type, ROW_NUMBER() OVER(
+                                PARTITION BY recipesymbol_id
+                                ORDER BY commit_date DESC, id DESC
+                            ) AS rownum
+                        FROM rrs_recipeupgrade
+                        WHERE commit_date <= %s
+                        AND upgrade_type not in ('M', 'N')) AS te
+                        INNER JOIN rrs_recipesymbol AS rs
+                        ON te.recipesymbol_id = rs.id
+                        WHERE rownum = 1
+                        AND te.upgrade_type <> 'R'
+                        AND rs.layerbranch_id = %s
+                        ORDER BY rs.pn;
+                        """, [date, layerbranch_id])
+        return Raw.dictfetchall(cur)
+
+    @staticmethod
+    def get_reupg_by_dates_and_recipes(start_date, end_date, recipes_id):
+        """  Get Recipe Upgrade for the milestone based on Recipes """
+        cur = connection.cursor()
+        qry = """SELECT DISTINCT recipesymbol_id
+                FROM rrs_recipeupgrade
+                WHERE commit_date >= %s
+                AND commit_date <= %s
+                AND recipesymbol_id IN %s;"""
+        cur.execute(qry, [start_date, end_date, tuple(recipes_id)])
+        return Raw.dictfetchall(cur)
+
+    @staticmethod
+    def get_remahi_by_end_date(layerbranch_id, date):
+        """ Get the latest Recipe Maintainer History for the milestone """
+        cur = connection.cursor()
+
+        cur.execute("""SELECT id
+                        FROM rrs_recipemaintainerhistory
+                        WHERE date <= %s
+                        AND layerbranch_id = %s
+                        ORDER BY date DESC
+                        LIMIT 1;
+                    """, [str(date), layerbranch_id])
+
+        ret = cur.fetchone()
+
+        if not ret:
+            cur.execute("""SELECT id
+                        FROM rrs_recipemaintainerhistory
+                        WHERE layerbranch_id = %s
+                        ORDER BY date
+                        LIMIT 1;
+                        """, [layerbranch_id])
+            ret = cur.fetchone()
+
+        return ret
+
+    @staticmethod
+    def dictfetchall(cursor):
+        "Returns all rows from a cursor as a dict"
+        desc = cursor.description
+        return [
+            dict(zip([col[0] for col in desc], row))
+            for row in cursor.fetchall()
+        ]
+
+
+def _get_milestone_statistics(milestone, maintainer_name=None):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 51 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 98% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code block on lines 344-371 into a separate function.